File: android/src/main/java/expo/modules/comergeruntime/ComergeRuntimeModule.kt
package expo.modules.comergeruntime

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ComergeRuntimeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ComergeRuntime")

    Events("onRuntimeMessage")

    Function("emitSystemEvent") { envelope: Map<String, Any?> ->
      if (!ComergeRuntimeBridgeValidation.isValidEnvelope(envelope, requireMicroSource = true)) {
        return@Function false
      }
      ComergeRuntimeBridge.dispatchMicroToHost(envelope)
    }

    Function("postMessageToRuntime") { envelope: Map<String, Any?> ->
      if (!ComergeRuntimeBridgeValidation.isValidEnvelope(envelope, requireMicroSource = false)) {
        return@Function false
      }
      ComergeRuntimeBridge.dispatchHostToMicro(envelope)
    }

    View(ComergeRuntimeView::class) {
      Events("onMessage")

      Prop("appKey") { view: ComergeRuntimeView, value: String? ->
        view.setAppKey(value)
      }

      Prop("bundlePath") { view: ComergeRuntimeView, value: String? ->
        view.setBundlePath(value)
      }

      Prop("runtimeId") { view: ComergeRuntimeView, value: String? ->
        view.setRuntimeId(value)
      }

      Prop("initialProps") { view: ComergeRuntimeView, value: Map<String, Any?>? ->
        view.setInitialProps(value)
      }
    }
  }
}

File: ios/ComergeRuntimeModule.swift
import ExpoModulesCore

public class ComergeRuntimeModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ComergeRuntime")

    Events("onRuntimeMessage")

    OnStartObserving("onRuntimeMessage") {
      NotificationCenter.default.addObserver(
        self,
        selector: #selector(handleHostMessageNotification),
        name: comergeRuntimeOnHostMessageNotification,
        object: nil
      )
    }

    OnStopObserving("onRuntimeMessage") {
      // swiftlint:disable:next notification_center_detachment
      NotificationCenter.default.removeObserver(self, name: comergeRuntimeOnHostMessageNotification, object: nil)
    }

    Function("emitSystemEvent") { (envelope: [String: Any]) -> Bool in
      guard ComergeRuntimeBridgeValidation.isValidEnvelope(envelope, requireMicroSource: true) else {
        return false
      }
      return ComergeRuntimeBridge.shared.dispatchMicroToHost(envelope: envelope)
    }

    Function("postMessageToRuntime") { (envelope: [String: Any]) -> Bool in
      guard ComergeRuntimeBridgeValidation.isValidEnvelope(envelope, requireMicroSource: false) else {
        return false
      }
      return ComergeRuntimeBridge.shared.dispatchHostToMicro(envelope: envelope)
    }

    View(ComergeRuntimeExpoView.self) {
      Events("onMessage")

      Prop("appKey") { (view: ComergeRuntimeExpoView, value: String?) in
        view.setAppKey(value)
      }

      Prop("bundlePath") { (view: ComergeRuntimeExpoView, value: String?) in
        view.setBundlePath(value)
      }

      Prop("runtimeId") { (view: ComergeRuntimeExpoView, value: String?) in
        view.setRuntimeId(value)
      }

      Prop("initialProps") { (view: ComergeRuntimeExpoView, value: [String: Any]?) in
        view.setInitialProps(value)
      }
    }
  }

  @objc private func handleHostMessageNotification(_ notification: Notification) {
    guard let envelope = notification.userInfo?["envelope"] as? [String: Any] else {
      return
    }
    sendEvent("onRuntimeMessage", ["envelope": envelope])
  }
}
